merkle/src/merkle.rs

```diff
@@ -586,72 +623,134 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
+                (leaves_lock.clone(), top_half_lock.clone(), 0, 0)
             } else {
-                panic!("MT build: inconsistent i: {} and j: {}", i, j);
+                // For all other levels we'll read/write from/to the `top_half` adjusting the
+                // "global" index to access this `Store` (offsetting `leaves` length). All levels
+                // are contiguous so we read/write `width` nodes apart.
+                let read_start = level_node_index - leaves_lock.read().unwrap().len();
+                (
+                    top_half_lock.clone(),
+                    top_half_lock.clone(),
+                    read_start,
+                    read_start + width,
+                )
             };
 
-            // TODO: avoid collecting into a vec and write the results direclty if possible.
-            for el in layer.into_iter() {
-                self.top_half.push(el);
-            }
-
+            // Allocate `width` indexes during operation (which is a negligible memory bloat
+            // compared to the 32-bytes size of the nodes stored in the `Store`s) and hash each
+            // pair of nodes to write them to the next level in concurrent threads.
+            // Process `chunk_size` nodes in each thread at a time to reduce contention, optimized
+            // for big sector sizes (small ones will just have one thread doing all the work).
+            let chunk_size = 1024;
+            debug_assert_eq!(chunk_size % 2, 0);
+            Vec::from_iter((read_start..read_start + width).step_by(chunk_size))
+                .par_chunks(1)
+                .for_each(|chunk_index| {
+                    let chunk_index = chunk_index[0];
+                    let chunk_size = std::cmp::min(chunk_size, read_start + width - chunk_index);
+
+                    let chunk_nodes = {
+                        // Read everything taking the lock once.
+                        let read_store = read_store_lock.read().unwrap();
+                        read_store.read_range(std::ops::Range {
+                            start: chunk_index,
+                            end: chunk_index + chunk_size,
+                        })
+                    };
+
+                    let hashed_nodes = chunk_nodes.chunks(2).map(|node_pair| {
+                        A::default().node(node_pair[0].clone(), node_pair[1].clone(), level)
+                        // FIXME: Change `node()` to receive references to avoid the `clone` here,
+                        //  this might be an API-breaking change.
+                    });
+
+                    // We write the hashed nodes to the next level in the position that
+                    // would be "in the middle" of the previous pair (dividing by 2).
+                    let write_delta = (chunk_index - read_start) / 2;
+
+                    let mut hashed_nodes_as_bytes: Vec<u8> =
+                        Vec::with_capacity(hashed_nodes.len() * T::byte_len());
```

Comment:
Can't we write directly to the store, rather than allocating a vector just to hold the conversions?

Reply:
Yes, but the priority here is to reduce lock contention. If we write directly to the store we need to write each `Element` at a time (converting it to a byte slice in the process) instead of doing a single `memcpy` operation (it's the price of the `Element` abstraction).

Reply:
Right… I think this gets back to questions around how to most-efficiently parallelize as well. I accept that we shouldn't worry about this here/now.